truffle/src/main/java/org/jruby/truffle/parser/MethodTranslator.java

```diff
@@ -449,4 +469,33 @@ private static String indentation(String line) {
         return "";
     }
 
+    /*
+     * The following methods allow us to save and restore enough of
+     * the current state of the Translator to allow lazy parsing. When
+     * the lazy parsing is actually performed, the state is restored
+     * to what it would have been if the method had been parsed
+     * eagerly.
+     */
+    public TranslatorState getCurrentState() {
+        return new TranslatorState(getEnvironment().unsafeGetLexicalScope(), getEnvironment().isDynamicConstantLookup(), new ArrayDeque<>(parentSourceSection));
```

Comment:
Yeah I'm just leaving the stack alone at the moment as I don't totally understand it. I'll look at it again later.